src/libPMacc/include/mappings/threads/ThreadCollective.hpp

```diff
@@ -32,72 +32,61 @@
- * @tparam T_numWorkers number of workers which are used to execute this functor
+ * @tparam T_DataDomain PMacc::SuperCellDescription, compile time data domain
+ *                      description with a CORE and GUARD
+ * @tparam T_numWorkers number of workers
  */
 template<
-    typename T_BlockDomain,
-    uint32_t T_numWorkers = math::CT::volume<typename T_BlockDomain::SuperCellSize>::type::value
+    typename T_DataDomain,
+    uint32_t T_numWorkers
 >
 class ThreadCollective
 {
 private:
-    /** size of the inner domain */
-    using SuperCellSize = typename T_BlockDomain::SuperCellSize;
+    // size of the CORE (in elements per dimension)
+    using CoreDomainSize = typename T_DataDomain::SuperCellSize;
+    // full size if the domain including the GUARD (in elements per dimension)
+    using DomainSize = typename T_DataDomain::FullSuperCellSize;
+    // offset (in elements per dimension) from the GUARD origin to the CORE
+    using OffsetOrigin = typename T_DataDomain::OffsetOrigin;
 
-    /** full size of the domain including the guards around SuperCellSize */
-    using FullSuperCellSize = typename T_BlockDomain::FullSuperCellSize;
-
-    /** size of the negative guard */
-    using OffsetOrigin = typename T_BlockDomain::OffsetOrigin;
-
-    /** number of worker performing the functior */
     static constexpr uint32_t numWorkers = T_numWorkers;
+    static constexpr uint32_t dim = T_DataDomain::Dim;
 
-    /** dimensionality of the index domain */
-    static constexpr uint32_t dim = T_BlockDomain::Dim;
-
-    /** index of the worker: range [0;numWorkers) */
-    PMACC_ALIGN( m_workerIdx, uint32_t const );
+    const PMACC_ALIGN(
+        m_workerIdx,
+        uint32_t
+    );
 
 public:
 
     /** constructor
      *
-     * @param workerIdx worker index
+     * @param workerIdx index of the worker
      */
     DINLINE ThreadCollective( uint32_t const workerIdx ) :
         m_workerIdx( workerIdx )
-    { }
-
-    /** constructor
-     *
-     * @warning this constructor is **deprecated** (assuming that the index domain
-     *          and the number of workers needs to be euqal is not good and avoid the
-     *          usage of alpaka)
-     *
-     * @param nDimWorkerIdx n dimensional worker index
-     */
-    DINLINE ThreadCollective( DataSpace< dim > const nDimWorkerIdx ) :
-        m_workerIdx( DataSpaceOperations< dim >::template map< SuperCellSize >( nDimWorkerIdx ) )
-    { }
+    {
+    }
 
-    /** execute a functor for each data point
+    /** execute the user functor for each element in the full domain
      *
-     * @tparam T_Functor type of the functor
-     * @tparam T_Args type of arguments given to the functor
+     * @tparam T_Functor type of the user functor, must have a `void operator()`
+     *                   with as many arguments as args contains
+     * @tapram  T_Args type of the arguments, each type must implement an operator
+     *                 `template<typename T, typnme R> R operator(T)`
```

Comment:
Wait, I thought @psychocoderHPC does at least compile and test run every change once. :open_mouth: